### PR TITLE
refactor(transport): lazy span-view TensorRef + find/move/replace localize

### PR DESCRIPTION
**Pull request by celve**
## Summary

Behavior-preserving refactor (plus one correctness fix) of the tensor **transport** layer in `unirl/distributed/tensor/` — the data plane between the FSDP trainer and the rollout engines for both AR and diffusion, so a regression here breaks every e2e run.

This **extends and supersedes the transport portion of #49**: it includes #49's selection-without-hydration foundation and builds the full ref/span model, the find/move/replace `localize`, and a module split on top. (#49's DP-seqlen-balance changes are separately tracked in #45 and are not part of this PR.)

- **Lazy span-view selection.** `TensorMeta`→`TensorRef`, `HandleView`→`TensorSpan` (now generic over a minimal `TensorHandle` protocol). A `TensorRef` is one `concat_field` over a uniform `List[TensorSpan]`; `select`/`slice`/`select_ranges` emit row-window views with **no data motion** — only `materialize` fetches. Six selection entry points collapse to one range engine (`select_ranges`, renamed from `select_segments` to avoid the `Segment`/`LatentSegment` domain collision).
- **`localize` = find / move / replace.** A pure FIND walk classifies each span local-vs-foreign per backend `_is_local`; MOVE runs one batched NCCL hop per `(src,dst)` device pair; REPLACE swaps foreign spans back. A no-op move now returns the **same object**, preserving `grad`/`retain_grad`/`_packed_cu_seqlens` that the old unconditional `with_spans` dropped — **correctness fix**. `routing_copy` and the `id()`-keyed substitution are removed (pickling already strips the worker binding, so shipping the original span is equivalent).
- **gpu_store resolve funnel + refcount-driven borrow GC.** Span slicing + `cat_rows` lift into the base `get`/`get_batch`; backends expose only `_resolve_handles`. The manual IPC-borrow bookkeeping (`_open_storages`/`_resolved_map`/`end_call`) is gone — a resolved tensor keeps its borrowed view alive via storage refcount, so the mapping closes itself when the last slice is dropped. Span-level NCCL routing ships only `[start:stop)` rows and allocates the sliced shape on recv.
- **`cat_rows`** right-pads ragged 2D+ parts with zeros to max width.
- **Module split (no behavior change).** The 937-line `transport.py` splits along its seams into `ref.py` (backend-agnostic data model), `transport.py` (ABC + session + runtime), and `worker_local.py` (`WorkerLocalTransport` + localize).

Net: 29 files, +952 / −713.

## Related Issue

Part of #40 (perf series). Supersedes the transport portion of #49.

## Test Plan

The transport pytest suite used to validate this is maintained out-of-tree and is **not** included in this PR; the PR keeps the tier markers (`pyproject.toml [tool.pytest.ini_options]`) and the resource-autodetecting runner (`scripts/run_transport_tests.sh`) so the suite drops in and runs.

Validation run on 1×8 H20:

- Full transport suite (CPU + gpu + multigpu tiers, incl. a 300-iter refcount/borrow no-leak loop): **185 passed, 4 skipped** (tq lib / Mooncake absent).
- Diffusion e2e — SD3, FlowGRPO + PickScore, 8×H20, LoRA — per-rollout PickScore grows **0.747 → 0.854** over 17 rollouts; both `colocate_store` and a `gpu_store` variant exercised under real NCCL.
- `ruff check` / `ruff format`: clean.

AR DRPO e2e was blocked before any tokens by an unrelated sglang chat-template/tokenizer bug (`apply_chat_template` returning a `tokenizers.Encoding`), orthogonal to transport; the AR transport path is covered by the multigpu NCCL move / on-wire dedup / grad tests.

## Compatibility / Risk

- Internal-only API renames within `unirl/distributed/tensor/`: `TensorMeta`→`TensorRef`, `HandleView`→`TensorSpan`, `select_segments`→`select_ranges`. All in-tree importers repointed. No public config, checkpoint, data-format, or CLI changes.
- Behavior change (a fix): a no-op `localize` returns the same object, so `grad`/`_packed_cu_seqlens` now survive it.
- Touches the trainer↔rollout data plane for both AR and diffusion; validated under real multi-GPU NCCL (see Test Plan).

## Reviewer Notes

- **Relationship to other PRs:** extends and supersedes the transport portion of #49 (`perf/07`); its DP-seqlen-balance changes live in #45 (`perf/04`) and are intentionally excluded here. Overlap check: no other open PR touches `unirl/distributed/tensor/`.
- Suggested reading order: `ref.py` (data model) → `worker_local.py` (`localize` find/move/replace) → `transport.py` (base resolve/get funnel) → backend deltas (`gpu_store`, `colocate_store`, `transfer_queue`).
- The kept `scripts/run_transport_tests.sh` + pytest markers are deliberate test infra; the suite files themselves are maintained out-of-tree per maintainer preference.
- Developed with AI assistance; the submitter has reviewed the full diff and run the validation above.

## Checklist

- [x] I reviewed the changed code and removed unrelated/generated artifacts.
- [x] I updated tests, docs, and configs where needed, or explained why not.

### CODE REVIEW

**Comment by leviking98z-rgb** [comment]
## Review follow-up: independent CPU validation (the dropped in-tree coverage)

Since the transport suite is kept out-of-tree, I ran an independent CPU validation against this PR's actual code (`f1286b6`) to exercise the part most amenable to unit testing — the pure data model (`ref.py`) and the `localize` no-op path. **23/23 pass, `ruff check` clean.**

```
$ git archive <pr> unirl | tar -x -C /tmp/src && PYTHONPATH=/tmp/src python pr59_validate.py
ALL PASS  (23 tests)
ruff check: All checks passed!
```

### Covered
| area | checks |
|---|---|
| `TensorSpan` | basic window, nested flatten (`TensorSpan(TensorSpan(h,2,8),1,3) == h[3:5]`), invalid-range raise, pickle round-trip |
| `cat_rows` | empty / single passthrough / same-width / **ragged right-pad to max width** |
| `select_ranges` | boundary-aligned **reuses the original span object** (zero-cost), partial wrap, **cross-span-boundary** (`[h(4),h(4)]` select `[2,6)` → `[2,3,0,1]`), out-of-bounds raise, empty |
| `select` (gather/permute) | materialize order matches `t[perm]`; consecutive run coalesces to one span |
| `slice` / `__getitem__` / `concat` / `with_spans` | slice == `t[a:b]`; `with_spans` drops grad (view default) |
| `map_tree` | `TensorRef` is an atomic leaf (spans never recursed), structural substitution |
| **`localize` no-op (this PR's fix)** | all-local shards return the **same object**; `_replace_leaf` returns the same ref and **grad survives** |
| `_move_key` | local → `None`; foreign → `(src_dev, dst_dev, store_key, start, stop)` |

The cross-span-boundary case is the one I most wanted to pin: selecting `[2,6)` over two 4-row handles correctly takes span0`[2:4)` + span1`[0:2)` and materializes `[2,3,0,1]` — the range engine splits at the seam without dropping or duplicating rows.

### Not covered here (needs GPU + multi-worker Ray — your out-of-tree suite owns these)
- the NCCL `_move` cross-worker hop
- the gpu_store IPC borrow + **refcount-driven GC** (worth a dedicated no-leak loop given it moves from explicit `end_call` to implicit refcount — a retained resolved-tensor/slice now pins producer memory)
- end-to-end reward curve

### Offer
This validation is a ~230-line pure-CPU file with no Ray/GPU deps — it could drop straight in as `tests/test_tensorref.py` and restore in-tree, CI-runnable coverage for the data model. Happy to push it onto this branch (or hand it over) if useful.

<details>
<summary><code>pr59_validate.py</code> — full source (the suite that produced the above)</summary>

```python
"""Standalone validation for PR #59 transport rewrite (ref.py + localize no-op).

Recreates the in-tree CPU coverage the PR dropped. Pure CPU — no Ray actors, no
GPU. Run: PYTHONPATH=<pr59 src> python pr59_validate.py
"""

from __future__ import annotations

import torch

from unirl.distributed.tensor.ref import TensorRef, TensorSpan, cat_rows, map_tree


class FakeHandle:
    """Satisfies the TensorHandle protocol + the localize-side attrs."""

    def __init__(self, t, source_id="w0", store_key=None, object_ref=None):
        self._t = t
        self.source_id = source_id
        self.store_key = store_key if store_key is not None else f"k{id(self)}"
        self.object_ref = object_ref

    def local(self):
        return self._t

    @property
    def shape(self):
        return tuple(self._t.shape)

    @property
    def dtype(self):
        return self._t.dtype

    @property
    def device(self):
        return str(self._t.device)


def _h(nrows, width=2, **kw):
    # rows are 0,1,2,... so we can assert exact gather order after materialize
    t = torch.arange(nrows, dtype=torch.float32).reshape(nrows, 1).repeat(1, width)
    return FakeHandle(t, **kw)


# ───────────────────────── TensorSpan ─────────────────────────
def test_span_basic():
    h = _h(10)
    s = TensorSpan(h, 1, 4)
    assert s.nrows == 3 and len(s) == 3
    assert s.shape == (3, 2)
    assert torch.equal(s.local(), h.local()[1:4])


def test_span_nested_flatten():
    h = _h(10)
    s = TensorSpan(TensorSpan(h, 2, 8), 1, 3)
    assert s.handle is h and (s.start, s.stop) == (3, 5)
    assert torch.equal(s.local(), h.local()[3:5])


def test_span_invalid_range():
    h = _h(10)
    try:
        TensorSpan(h, 5, 2)
    except ValueError:
        return
    raise AssertionError("expected ValueError for stop < start")


def test_span_pickle_roundtrip():
    import pickle

    h = _h(6)
    s = TensorSpan(h, 1, 4)
    s2 = pickle.loads(pickle.dumps(s))
    assert (s2.start, s2.stop) == (1, 4) and torch.equal(s2.local(), s.local())


# ───────────────────────── cat_rows ─────────────────────────
def test_cat_rows_empty_single():
    assert cat_rows([]).numel() == 0
    t = torch.ones(3, 2)
    assert cat_rows([t]) is t  # single part passes through


def test_cat_rows_same_width():
    out = cat_rows([torch.ones(2, 3), torch.ones(1, 3)])
    assert out.shape == (3, 3)


def test_cat_rows_ragged_right_pad():
    a = torch.ones(2, 2)
    b = torch.ones(1, 4)
    out = cat_rows([a, b])
    assert out.shape == (3, 4)
    assert torch.equal(out[:2, 2:], torch.zeros(2, 2))  # a right-padded with zeros
    assert torch.equal(out[2], torch.ones(4))


# ───────────────────────── TensorRef select / ranges ─────────────────────────
def test_from_handles_sizes():
    r = TensorRef.from_handles([_h(4), _h(3)])
    assert r.sizes == [4, 3] and r.batch_size == 7 and r.shape == (7, 2)
    assert len(r.spans) == 2


def test_select_ranges_aligned_passthrough():
    h1, h2 = _h(4), _h(4)
    r = TensorRef.from_handles([h1, h2])
    s0 = r.spans[0]
    out = r.select_ranges([(0, 4)])  # exactly span 0
    assert out.spans[0] is s0  # zero-cost: original span object reused


def test_select_ranges_partial_wrap():
    r = TensorRef.from_handles([_h(8)])
    out = r.select_ranges([(2, 5)])
    assert out.batch_size == 3
    assert torch.equal(out.materialize(), torch.arange(2, 5).reshape(3, 1).repeat(1, 2).float())


def test_select_ranges_crosses_span_boundary():
    r = TensorRef.from_handles([_h(4), _h(4)])  # rows 0-3 then 0-3
    out = r.select_ranges([(2, 6)])  # last 2 of span0 + first 2 of span1
    got = out.materialize()
    exp = torch.tensor([2.0, 3.0, 0.0, 1.0]).reshape(4, 1).repeat(1, 2)
    assert torch.equal(got, exp)


def test_select_ranges_out_of_bounds():
    r = TensorRef.from_handles([_h(4)])
    try:
        r.select_ranges([(0, 5)])
    except IndexError:
        return
    raise AssertionError("expected IndexError for range past end")


def test_select_ranges_empty():
    r = TensorRef.from_handles([_h(4)])
    out = r.select_ranges([])
    assert out.batch_size == 0 and out.spans == []


def test_select_permute_materialize():
    r = TensorRef.from_handles([_h(6)])
    perm = [3, 1, 4, 1, 5]
    out = r.select(perm)
    exp = r.materialize()[perm]
    assert torch.equal(out.materialize(), exp)


def test_select_consecutive_coalesces():
    r = TensorRef.from_handles([_h(6)])
    out = r.select([2, 3, 4])
    assert len(out.spans) == 1  # consecutive run → one span
    assert torch.equal(out.materialize(), r.materialize()[2:5])


def test_slice_and_getitem():
    r = TensorRef.from_handles([_h(6)])
    assert torch.equal(r.slice(1, 4).materialize(), r.materialize()[1:4])
    assert torch.equal(r[1:4].materialize(), r.materialize()[1:4])


def test_concat():
    a = TensorRef.from_handles([_h(3)])
    b = TensorRef.from_handles([_h(2)])
    c = TensorRef.concat([a, b])
    assert c.batch_size == 5 and len(c.spans) == 2


def test_with_spans_drops_grad():
    r = TensorRef.from_handles([_h(3)])
    r.grad = TensorRef.from_handles([_h(3)])
    r.retain_grad_flag = True
    r2 = r.with_spans(r.spans)
    assert r2.grad is None and r2.retain_grad_flag is False  # view defaults


# ───────────────────────── map_tree ─────────────────────────
def test_map_tree_ref_is_atomic():
    r = TensorRef.from_handles([_h(3)])
    seen = []

    def leaf(o):
        seen.append(type(o).__name__)
        return o

    out = map_tree({"x": [r, 5], "y": (r,)}, leaf)
    assert out["x"][1] == 5
    # the ref appears as a leaf and is NOT recursed into (spans not visited)
    assert "TensorSpan" not in seen and seen.count("TensorRef") == 2


def test_map_tree_substitution():
    r = TensorRef.from_handles([_h(3)])
    r2 = TensorRef.from_handles([_h(3)])

    def leaf(o):
        return r2 if o is r else o

    out = map_tree([r, {"k": r}], leaf)
    assert out[0] is r2 and out[1]["k"] is r2


# ───────────────────────── localize no-op (the PR's bug fix) ─────────────────────────
def test_localize_noop_returns_same_object():
    from unirl.distributed.tensor.worker_local import WorkerLocalTransport

    class FakePool:
        def device_id_of(self, sid):
            return 0

    h = _h(4, source_id="w0")
    ref = TensorRef.from_handles([h])
    shards = [((ref,), {})]
    out = WorkerLocalTransport.localize(shards, FakePool(), device_ids=[0], worker_ids=["w0"])
    assert out is shards  # all-local → untouched same object


def test_replace_leaf_preserves_same_ref_and_grad():
    from unirl.distributed.tensor.worker_local import WorkerLocalTransport

    class FakePool:
        def device_id_of(self, sid):
            return 0

    h = _h(4, source_id="w0")
    ref = TensorRef.from_handles([h])
    ref.grad = TensorRef.from_handles([_h(4)])
    leaf = WorkerLocalTransport._replace_leaf(moved={}, dst=("w0", 0), pool=FakePool())
    out = leaf(ref)
    assert out is ref and out.grad is ref.grad  # unchanged → same object, grad survives


def test_move_key_foreign_vs_local():
    from unirl.distributed.tensor.worker_local import WorkerLocalTransport

    class FakePool:
        def device_id_of(self, sid):
            return {"w0": 0, "w1": 1}[sid]

    local = TensorSpan(_h(4, source_id="w0"), 0, 4)
    foreign = TensorSpan(_h(4, source_id="w1", store_key="kf"), 0, 4)
    assert WorkerLocalTransport._move_key(local, ("w0", 0), FakePool()) is None
    k = WorkerLocalTransport._move_key(foreign, ("w0", 0), FakePool())
    assert k == (1, 0, "kf", 0, 4)  # (src_dev, dst_dev, store_key, start, stop)


if __name__ == "__main__":
    fails = 0
    for name, fn in sorted(globals().items()):
        if not name.startswith("test_") or not callable(fn):
            continue
        try:
            fn()
            print(f"PASS {name}")
        except Exception as exc:  # noqa: BLE001
            fails += 1
            import traceback

            print(f"FAIL {name}: {type(exc).__name__}: {exc}")
            traceback.print_exc()
    print(f"\n{'ALL PASS' if not fails else f'{fails} FAILURE(S)'}")
    raise SystemExit(1 if fails else 0)

```

</details>

**Review by leviking98z-rgb — APPROVED** [review]
